### ISSUE TITLE
Comments about the timeout handling.

### ISSUE THREAD

**Issue by kery** [shadowsocks/shadowsocks-libev#2251]
### What version of shadowsocks-libev are you using?
v3.2.3

### What operating system are you using?
Linux, Cygwin

### What did you do?
Not important

### What did you expect to see?
Maybe it is better to make SS to support long time idle connection, by setting the timeout value (e.g. 0). Sometimes, we need to keep the TCP connection even there is no traffic between SS client and server. SS is not only used for web traffic, it also useful in other scenarios. So, it would be good if user can have more control of the timeout parameter.

In addition, currently both the clients (local, tunnel, etc.) and the server have timeout mechanism, maybe it's redundant, only keep the server side is enough (?).

### What did you see instead?
TCP connection is closed by SS when I do nothing for a while on my vnc viewer window.

### What is your config in detail (with all sensitive info masked)?
Not important

**Comment by madeye** [comment]
Just enlarge the timeout settings on both server and client sides.

**Comment by kery** [comment]
Yes, it can work by setting a large enough timeout value on both client and server sides, but this not perfectly meet the requirement. How do you think about making both the client and server to support for disabling the timeout feature by setting timeout value to a zero or negative value?

**Comment by madeye** [comment]
Nope, and you shouldn't do that, especially for server side.

Setting a large enough timeout already meets your requirements.

**Comment by kery** [comment]
Hi, I have another question about the timeout handling. In the current implementation, if there is no downlink traffic (server->client) and the timeout value elapsed, then the connection will be closed even if there is uplink traffic. Was it deliberately designed like this?

Thanks!

**Comment by madeye** [comment]
When any byte sent

from ss-local to ss-server
https://github.com/shadowsocks/shadowsocks-libev/blob/9545ce20a41865249d86e2c05c24d25c01250eb0/src/server.c#L727

from remote server to ss-server
https://github.com/shadowsocks/shadowsocks-libev/blob/9545ce20a41865249d86e2c05c24d25c01250eb0/src/server.c#L1140

from local client to ss-local
https://github.com/shadowsocks/shadowsocks-libev/blob/9545ce20a41865249d86e2c05c24d25c01250eb0/src/local.c#L645

from ss-server to ss-local
https://github.com/shadowsocks/shadowsocks-libev/blob/9545ce20a41865249d86e2c05c24d25c01250eb0/src/local.c#L1012

the timeout timer will be reset.

So, if there is any uplink traffic, the connection won't be closed by shadowsocks.

I suggest you check with your ISP and cloud provider, to make sure their NAT device didn't close your connection.
